[ExternalLibrary] Adding csv-parser

loumalouomega opened the pull request:
This adds the library csv-parser (https://github.com/vincentlaucsb/csv-parser) in order to process CSV files, which are extensively used as input and output in many applications. I chose this one as it is C++11 compatible (I found some more interesting, but C++17> only) and it is MIT license.

As described in the original project:

 > **Motivation**

> There's plenty of other CSV parsers in the wild, but I had a hard time finding what I wanted. Inspired by Python's csv module, I wanted a library with simple, intuitive syntax. Furthermore, I wanted support for special use cases such as calculating statistics on very large files. Thus, this library was created with these following goals in mind.
Performance

> This CSV parser uses multiple threads to simulatenously pull data from disk and parse it. Furthermore, it is capable of incremental streaming (parsing larger than RAM files), and quickly parsing data types.

Additionally includes some interesting features as a JSON conversion output, so it could be parsed into a Parameters.

I adapted all the tests from the original library to the Kratos format. I am planning to add a CSVUtilities, but I don't know if do it in this PR or in a future PR after this is merged.

loumalouomega committed:
Adding csv-parser

loumalouomega committed:
Move

loumalouomega committed:
Missing CMake

loumalouomega committed:
Adding to cmake

loumalouomega committed:
First test draft

loumalouomega committed:
Update README

loumalouomega committed:
Update

loumalouomega committed:
Clean up

loumalouomega committed:
Linking

loumalouomega committed:
Extend

loumalouomega committed:
COrrect

loumalouomega committed:
Avoid -

loumalouomega committed:
Clean up

loumalouomega committed:
More test adapting

loumalouomega committed:
More update

loumalouomega committed:
Some revert

loumalouomega committed:
Update

loumalouomega committed:
Minor

loumalouomega committed:
Activating test (I)

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega committed:
Activating (II)

loumalouomega committed:
Activating test (III)

loumalouomega committed:
Activating test (IV)

loumalouomega committed:
Activate test (V)

loumalouomega committed:
Activating test (VI-a)

loumalouomega committed:
Activating test (VI-b)

loumalouomega committed:
Activating test (VII)

loumalouomega committed:
Activating test (VIII)

loumalouomega committed:
Minor

loumalouomega committed:
Activating test (IX)

loumalouomega committed:
Activate test (X)

loumalouomega committed:
Activate test (XI)

loumalouomega committed:
Adding ErasePartialString

loumalouomega committed:
Clear (I)

loumalouomega committed:
Clear (II)

loumalouomega committed:
Clear (III)

loumalouomega committed:
Clear (IV)

loumalouomega committed:
Activate test (XII-a)

loumalouomega committed:
RemoveWhiteSpaces

loumalouomega committed:
Fix test

loumalouomega committed:
Minor

loumalouomega committed:
Clean up

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega committed:
Comment

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega committed:
Revert "Test"

This reverts commit 1f61f394a3747187a089ceeedc166e480fc527f5.

loumalouomega committed:
Test

loumalouomega committed:
Test 2

loumalouomega committed:
Revert "Test 2"

This reverts commit 3b7b4032376a0b7b5b3e9bbf6cb311cfd97db523.

loumalouomega committed:
Revert "New test"

This reverts commit b24d5f0c47483ec1afab3dd6db762cedb3089bff.

loumalouomega committed:
New test

loumalouomega committed:
One more test

loumalouomega committed:
Revert "One more test"

This reverts commit 302dfd885423e28b933af2fd7c93625eddb1631d.

loumalouomega committed:
Compilation deactivated by default

loumalouomega committed:
Setting on CMakeLists.txt

loumalouomega committed:
Unify with master

philbucher commented:
why do you add the tests of that library?
They are huge and usually we don't do that

loumalouomega commented:
> why do you add the tests of that library?
> They are huge and usually we don't do that

You can see that the tests are deactivated by default, they are compiled only if specified in the configure.sh. I add the tests as case examples of use

philbucher commented:
hm
how big are the files you are adding?

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega commented:
Library is 230 kB and the tests are 94 MB, I will remove as much tests and files as I could

loumalouomega committed:
Removing large files

loumalouomega commented:
> Library is 230 kB and the tests are 94 MB, I will remove as much tests and files as I could

Now it is 4.8 MB for tests

philbucher commented:
But why do we need huge csv files for showing how it works? For this also very small files are sufficient

in any case, writing a csv-file is not rocket-sience...

BTW this PR needs to be squashed when merging, otherwise we will have the huge files in the git-history (which is already 10x the size of the repo itself)

loumalouomega committed:
Even more file cleaning

loumalouomega commented:
> squashed

How do I squash?

philbucher commented:
> > squashed
> 
> How do I squash?

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github

loumalouomega commented:
> > > squashed
> > 
> > 
> > How do I squash?
> 
> [help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github)

Now I see

RiccardoRossi commented:
doesn't CSV input come included with python? I don't love the idea of including another external library in our compilation...(not blocking, only discussing)

loumalouomega commented:
> doesn't CSV input come included with python? I don't love the idea of including another external library in our compilation...(not blocking, only discussing)

Well, but it that case you cannot access/generate CSV from C++ side. There is json support in python but we still have a library for json. Imagine you want to generate a CSV from the convergence criteria information. This should be done with a logger enhanced with the csv-parser

Commit:
Merge branch 'master' into external_lib/csv-parser

Commit:
Single header

philbucher commented on CMakeLists.txt:
```diff
@@ -56,6 +56,12 @@ message("-- No KRATOS_BUILD_TESTING is defined. Setting to ON")
+message("-- No KRATOS_BUILD_CSV_TESTING is defined. Setting to OFF. This is due to the fact that it is an external library and represents the need of compiling more than 60 tests. In order to reduce compilation time by default will not be compiled")
+SET (KRATOS_BUILD_CSV_TESTING OFF)
+endif(NOT DEFINED KRATOS_BUILD_CSV_TESTING)
+
```
this was removed no?

loumalouomega replied:
Tests (some of them) are still there. I can remove them, but these are useful in order to see how to use the library, and by default are not build

philbucher commented on external_libraries/csv-parser/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.9)
```
this file seems to be leftover

loumalouomega replied:
Ups

Commit:
Merge branch 'master' into external_lib/csv-parser

Commit:
Leftover

Commit:
Update

Commit:
More leftover

philbucher commented on CMakeLists.txt:
```diff
@@ -56,6 +56,12 @@ message("-- No KRATOS_BUILD_TESTING is defined. Setting to ON")
+message("-- No KRATOS_BUILD_CSV_TESTING is defined. Setting to OFF. This is due to the fact that it is an external library and represents the need of compiling more than 60 tests. In order to reduce compilation time by default will not be compiled")
+SET (KRATOS_BUILD_CSV_TESTING OFF)
+endif(NOT DEFINED KRATOS_BUILD_CSV_TESTING)
+
```
```suggestion
```

can you please remove this, as we are fully integrating it we also want it tested properly

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_read_csv_file.cpp:
```diff
@@ -0,0 +1,184 @@
+//     const std::string working_dir = StringUtilities::ErasePartialString(__FILE__, "test_read_csv_file.cpp");
+//     int pos = get_col_pos(Kratos::FilesystemExtensions::JoinPaths({working_dir, "data/real_data/2015_StateDepartment.csv"}),"Entity Type");
+//     KRATOS_CHECK_EQUAL(pos, 1);
+// }
```
please remove commented code

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_write_csv.cpp:
```diff
@@ -0,0 +1,117 @@
+using namespace csv;
+using std::queue;
+using std::vector;
+using std::string;
```
why do you do this?
=> I would remove it as it makes the code less readable and others might use it then too
furthermore vector now collides with ublas vector!

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_write_csv.cpp:
```diff
@@ -0,0 +1,117 @@
+#ifdef KRATOS_BUILD_CSV_TESTING
+
+// System includes
+#include <stdio.h> // For remove()
```
where do you use this?

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_write_csv.cpp:
```diff
@@ -0,0 +1,117 @@
+//
+//
+
+#ifdef KRATOS_BUILD_CSV_TESTING
```
can be removed

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/data/fake_data/README.md:
```diff
@@ -0,0 +1,9 @@
+ints.csv
+ * Header row with columns "A", "B", ..., "J"
+ * 10 columns; each with the first 100 integers
+ * "\r\n" separates records
```
this file is not needed I would say

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_csv_buffer.cpp:
```diff
@@ -0,0 +1,126 @@
+
+// Project includes
+#include "testing/testing.h"
+#include "csv-parser/include/csv.hpp"
```
can you move the file so that the include path is shorter?
=> `#include "csv-parser/csv.hpp"`

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_csv_buffer.cpp:
```diff
@@ -0,0 +1,126 @@
+
+KRATOS_TEST_CASE_IN_SUITE(GiantStringBufferTest, KratosExternalLibrariesFastSuite)
+{
+    BufferPtr buffer = BufferPtr(new RawRowBuffer());
```
Where is this memory deleted?

philbucher commented on kratos/tests/cpp_tests/external_libraries/csv-parser/test_csv_buffer.cpp:
```diff
@@ -0,0 +1,126 @@
+//
+//
+
+#ifdef KRATOS_BUILD_CSV_TESTING
```
```suggestion
```

philbucher reviewed (commented):
some more comments related to cleanup

What I wrote for individual tests also is valid for other tests

Commit:
Merge branch 'master' into external_lib/csv-parser

Commit:
Remove legacy test (buffer removed in version 2.0)

Commit:
Minor fix some compilers

Commit:
Compiling always

Commit:
Remove commented code

Commit:
Missing macro

Commit:
Avoid using

Commit:
Remove test with large data (69 MB)

Commit:
Unused include

Commit:
Remove readme

Commit:
Merge branch 'master' into external_lib/csv-parser

Commit:
Remove operator

Commit:
C++17 only changed to C++11

Commit:
Fix

Commit:
Avoid asserts from C++17

Commit:
Revert "Avoid asserts from C++17"

This reverts commit 8288574406bec2d649b7524470757c1d625a5810.

Commit:
Parsing

philbucher committed:
fixing static_asserts

philbucher committed:
shortening include paths

philbucher committed:
move file

philbucher committed:
Merge remote-tracking branch 'origin/master' into external_lib/csv-parser

loumalouomega committed:
Merge branch 'master' into external_lib/csv-parser

loumalouomega committed:
Update exception

loumalouomega committed:
Pedantic

philbucher committed:
shouldn't be necessary any more

philbucher committed:
Revert "shouldn't be necessary any more"

This reverts commit 42a8343a8c2eb880f2339000149d4f40314cdfd9.

philbucher reviewed (commented):
@loumalouomega this should probably be redefined to nothing instead of disabling the warning

EDIT I am refering to my lasz commit

Commit:
Revert "[Core][Logger] Replace GetMessage with GetLoggerMessage to avoid conflict"

This reverts commit ac02f0b7bddb431fa1d536f8321cf80c8c2e30fe.

Commit:
[Core][Logger] Replace GetMessage with GetLoggerMessage to avoid conflict

Avoids conflict with Win32 Get;essage function: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage

Commit:
[Core][Logger] Replace GetMessage with GetLoggerMessage to avoid conflict

Avoids conflict with Win32 Get;essage function: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage

philbucher commented on external_libraries/csv-parser/csv.hpp:
```diff
@@ -4395,6 +4395,7 @@ HEDLEY_DIAGNOSTIC_POP
 #  if \
      (__cplusplus >= 201703L) || \
      ((__cplusplus >= 201103L) && HEDLEY_HAS_CPP_ATTRIBUTE(fallthrough))
+#pragma clang diagnostic ignored "-Wc++1z-extensions"
```
hm ok
Strange that they enable C++17 features ...

in any case, please also disable the warning removal so that it doesn't leak into other files

loumalouomega replied:
Is a pragma, I think it only affects to this

philbucher replied:
it does: https://clang.llvm.org/docs/UsersManual.html#id28

philbucher replied:
sorry thats the correct link: https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas

philbucher committed:
see what happens if windows.h is removed

philbucher committed:
see what happens if windows.h is removed

philbucher committed:
Revert "see what happens if windows.h is removed"

This reverts commit 5517435be0a81b66deeca4730db02cc33e0138ab.

roigcarlo committed:
Trying to chop some windows.h internals

roigcarlo committed:
Trying to chop some windows.h internals

roigcarlo committed:
Revert "Trying to chop some windows.h internals"

This reverts commit 4561a6005a4fdf405561818845bcc907f95f2b17.

Commit:
Merge branch 'master' into external_lib/csv-parser
